File: Sources/GUIForCLIWindows.Core/SetupModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace GUIForCLIWindows.Core;

public sealed record SetupSpec
{
    [JsonPropertyName("steps")]
    public List<SetupStepSpec> Steps { get; init; } = [];
}

public sealed record SetupStepSpec
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = "";

    [JsonPropertyName("label")]
    public string Label { get; init; } = "";

    [JsonPropertyName("kind")]
    public string Kind { get; init; } = "";

    [JsonPropertyName("command")]
    public CommandSpec? Command { get; init; }

    [JsonPropertyName("value")]
    public string? Value { get; init; }

    [JsonPropertyName("arguments")]
    public List<string> Arguments { get; init; } = [];

    [JsonPropertyName("environment")]
    public Dictionary<string, string> Environment { get; init; } = [];

    [JsonPropertyName("workingDirectory")]
    public string? WorkingDirectory { get; init; }

    [JsonPropertyName("optional")]
    public bool Optional { get; init; }

    [JsonPropertyName("packageId")]
    public string? PackageId { get; init; }

    [JsonPropertyName("script")]
    public string? Script { get; init; }
}

public sealed record ExitCodeReferenceSpec
{
    [JsonPropertyName("code")]
    public int Code { get; init; }

    [JsonPropertyName("title")]
    public string Title { get; init; } = "";

    [JsonPropertyName("summary")]
    public string Summary { get; init; } = "";

    [JsonPropertyName("severity")]
    public string Severity { get; init; } = "error";
}

File: Apps/Flutter/lib/src/models/setup_models.dart
part of '../models.dart';

class CommandSpec {
  const CommandSpec({
    required this.executable,
    required this.arguments,
    this.optionalArguments = const [],
  });

  final String executable;
  final List<String> arguments;
  final List<List<String>> optionalArguments;

  factory CommandSpec.fromJson(Map<String, Object?> json) => CommandSpec(
        executable: stringValue(json['executable']),
        arguments: listOfStrings(json['arguments']),
        optionalArguments: listOfStringLists(json['optionalArguments']),
      );
}

class SetupSpec {
  const SetupSpec({required this.steps});

  final List<SetupStepSpec> steps;

  factory SetupSpec.fromJson(Map<String, Object?> json) => SetupSpec(
        steps: listOfMaps(json['steps']).map(SetupStepSpec.fromJson).toList(),
      );
}

class SetupStepSpec {
  SetupStepSpec({
    required this.id,
    required this.label,
    required this.kind,
    this.value,
    this.workingDirectory,
    this.arguments = const [],
    this.environment = const {},
    this.optional = false,
  });

  final String id;
  final String label;
  final String kind;
  final String? value;
  final String? workingDirectory;
  final List<String> arguments;
  final Map<String, String> environment;
  final bool optional;

  factory SetupStepSpec.fromJson(Map<String, Object?> json) => SetupStepSpec(
        id: stringValue(json['id']),
        label: stringValue(json['label']),
        kind: stringValue(json['kind']),
        value: optionalString(json['value']),
        workingDirectory: optionalString(json['workingDirectory']),
        arguments: listOfStrings(json['arguments']),
        environment: mapOfStrings(json['environment']),
        optional: json['optional'] == true,
      );
}
